YouTube直播订阅报错

B站直播间添加正常，添加YouTube直播间之后，显示
[2021-04-09 04:41:20][Mirai] ERROR: Event: Error while fetching channel information: UCqm3BQLlJfvkTsX_hvm0UmA
    File "/root/MCL/MiraiBot/plugins/live_monitor/channels/youtube.py", line 42
  TypeError: 'NoneType' object is not subscriptable
[2021-04-09 04:41:40][Mirai] ERROR: Event: Error while fetching channel information: UC1opHUrw8rvnsadT-iGp7Cg
    File "/root/MCL/MiraiBot/plugins/live_monitor/channels/youtube.py", line 42
  TypeError: 'NoneType' object is not subscriptable
[2021-04-09 04:42:00][Mirai] ERROR: Event: Error while fetching channel information: UCQ0UDLQCjY0rmuxCDE38FGg
    File "/root/MCL/MiraiBot/plugins/live_monitor/channels/youtube.py", line 32
  TypeError: 'NoneType' object is not subscriptable

Bot正常回复“已添加1个频道”

[2021-04-10 10:31:37][Mirai] WARNING: Event: Fetching channel information failed: UC1opHUrw8rvnsadT-iGp7Cg
 [youtube.py][53] AssertionError: 获取直播间信息失败，可能处于未开播状态
[2021-04-10 10:31:57][Mirai] WARNING: Event: Fetching channel information failed: Aviation ZH
 [youtube.py][53] AssertionError: 获取直播间信息失败，可能处于未开播状态
[2021-04-10 10:32:57][Mirai] WARNING: Event: Fetching channel information failed: UC1opHUrw8rvnsadT-iGp7Cg
 [youtube.py][53] AssertionError: 获取直播间信息失败，可能处于未开播状态
[2021-04-10 10:33:17][Mirai] WARNING: Event: Fetching channel information failed: Aviation ZH
 [youtube.py][53] AssertionError: 获取直播间信息失败，可能处于未开播状态
[2021-04-10 10:34:17][Mirai] WARNING: Event: Fetching channel information failed: UC1opHUrw8rvnsadT-iGp7Cg
 [youtube.py][53] AssertionError: 获取直播间信息失败，可能处于未开播状态
[2021-04-10 10:34:37][Mirai] WARNING: Event: Fetching channel information failed: Aviation ZH
 [youtube.py][53] AssertionError: 获取直播间信息失败，可能处于未开播状态
[2021-04-10 10:35:37][Mirai] WARNING: Event: Fetching channel information failed: UC1opHUrw8rvnsadT-iGp7Cg
 [youtube.py][53] AssertionError: 获取直播间信息失败，可能处于未开播状态
[2021-04-10 10:36:57][Mirai] WARNING: Event: Fetching channel information failed: UC1opHUrw8rvnsadT-iGp7Cg
 [youtube.py][53] AssertionError: 获取直播间信息失败，可能处于未开播状态
[2021-04-10 10:38:17][Mirai] WARNING: Event: Fetching channel information failed: UC1opHUrw8rvnsadT-iGp7Cg
 [youtube.py][53] AssertionError: 获取直播间信息失败，可能处于未开播状态
[2021-04-10 10:39:37][Mirai] WARNING: Event: Fetching channel information failed: UC1opHUrw8rvnsadT-iGp7Cg
 [youtube.py][53] AssertionError: 获取直播间信息失败，可能处于未开播状态

我用Aviation ZH这个号在35分开始直播的

40分下播了log中又出现了

[2021-04-10 10:40:57][Mirai] WARNING: Event: Fetching channel information failed: UC1opHUrw8rvnsadT-iGp7Cg
 [youtube.py][53] AssertionError: 获取直播间信息失败，可能处于未开播状态
[2021-04-10 10:41:17][Mirai] WARNING: Event: Fetching channel information failed: Aviation ZH
 [youtube.py][53] AssertionError: 获取直播间信息失败，可能处于未开播状态
[2021-04-10 10:42:17][Mirai] WARNING: Event: Fetching channel information failed: UC1opHUrw8rvnsadT-iGp7Cg
 [youtube.py][53] AssertionError: 获取直播间信息失败，可能处于未开播状态
[2021-04-10 10:42:37][Mirai] WARNING: Event: Fetching channel information failed: Aviation ZH
 [youtube.py][53] AssertionError: 获取直播间信息失败，可能处于未开播状态
[2021-04-10 10:43:38][Mirai] WARNING: Event: Fetching channel information failed: UC1opHUrw8rvnsadT-iGp7Cg
 [youtube.py][53] AssertionError: 获取直播间信息失败，可能处于未开播状态
[2021-04-10 10:43:57][Mirai] WARNING: Event: Fetching channel information failed: Aviation ZH
 [youtube.py][53] AssertionError: 获取直播间信息失败，可能处于未开播状态
[2021-04-10 10:44:57][Mirai] WARNING: Event: Fetching channel information failed: UC1opHUrw8rvnsadT-iGp7Cg
 [youtube.py][53] AssertionError: 获取直播间信息失败，可能处于未开播状态
[2021-04-10 10:45:17][Mirai] WARNING: Event: Fetching channel information failed: Aviation ZH
 [youtube.py][53] AssertionError: 获取直播间信息失败，可能处于未开播状态
[2021-04-10 10:46:17][Mirai] WARNING: Event: Fetching channel information failed: UC1opHUrw8rvnsadT-iGp7Cg
 [youtube.py][53] AssertionError: 获取直播间信息失败，可能处于未开播状态
[2021-04-10 10:46:37][Mirai] WARNING: Event: Fetching channel information failed: Aviation ZH
 [youtube.py][53] AssertionError: 获取直播间信息失败，可能处于未开播状态

移除了B站的监控后测试，问题依旧

另外一个奇怪的地方：现在是开启了四个频道的监控（B站的监控移除了）
直播监控列表：第1页/共1页
[youtube_live]
Matsuri Channel 夏色まつり
Watame Ch. 角巻わため
UC1opHUrw8rvnsadT-iGp7Cg
Aviation ZH
但是log里面只出现了两位，而且另外两位我去看了一下直播没有开

--- a/plugins/live_monitor/channels/youtube.py
+++ b/plugins/live_monitor/channels/youtube.py
@@ -13,9 +13,72 @@ class YoutubeChannel(BaseChannel):
         return f'https://www.youtube.com/channel/{self.cid}/live'
 
     async def resolve(self, content: str):
-        tree: lxml.html.HtmlElement = lxml.html.fromstring(content)
-        script: str
-        if script := ''.join(tree.xpath('body/script[contains(text(),"RELATED_PLAYER_ARGS")]/text()')):
+        if '"isLive":true' not in content and self.cid in content:
+            return LiveCheckResponse(
+                name=self.ch_name,
+                live_status=0,
+                title='',
+                url=self.api_url,
+                cover=None
+            )
+
+        if resp := self.parse_html_1(content):
+            return resp
+        elif resp := self.parse_html_2(content):
+            return resp
+        elif resp := self.parse_html_3(content):
+            return resp
+        elif resp := self.parse_html_4(content):
+            return resp
+        else:
+            # ========== DEBUG ==========
+            import os
+            from config import data_path
+            debug_filepath = os.path.join(data_path, 'youtube_live_debug.html')
+            with open(debug_filepath, 'w', encoding='utf8') as f:
+                f.write(content)
+            # ========== DEBUG ==========
+
+            raise AssertionError(f'获取直播间信息失败, html页面保存至{debug_filepath}')
+
+    def parse_html_1(self, content: str):
+        if 'videoDetails' in content:
+            try:
+                tree: lxml.html.HtmlElement = lxml.html.fromstring(content)
+                script = ''.join(tree.xpath('body/script[contains(text(),"ytInitialPlayerResponse")]/text()'))
+                script = re.search(r'var ytInitialPlayerResponse = ({.*});', script)[1]
+                videoDetails = json.loads(script)['videoDetails']
+                return self.parse_videoDetails(videoDetails)
+            except:
+                return None
+
+    def parse_html_2(self, content: str):
+        if 'videoDetails' in content:
+            try:
+                tree: lxml.html.HtmlElement = lxml.html.fromstring(content)
+                script = ''.join(tree.xpath('body/script[contains(text(),"ytInitialPlayerResponse")]/text()'))
+                script = re.search(r'window\["ytInitialPlayerResponse"] = ({.*?});', script)[1]
+                videoDetails = json.loads(script)['videoDetails']
+                return self.parse_videoDetails(videoDetails)
+            except:
+                return None
+
+    def parse_html_3(self, content: str):
+        if 'videoDetails' in content:
+            try:
+                tree: lxml.html.HtmlElement = lxml.html.fromstring(content)
+                script = ''.join(tree.xpath('//div[@id="player-wrap"]/script[contains(text(),"player_response")]/text()'))
+                script = re.search(r'ytplayer.config = ({.*?});', script)[1]
+                script = json.loads(script)['args']['player_response']
+                videoDetails: Dict[str, Any] = json.loads(script)['videoDetails']
+                return self.parse_videoDetails(videoDetails)
+            except:
+                return None
+
+    def parse_html_4(self, content: str):
+        try:
+            tree: lxml.html.HtmlElement = lxml.html.fromstring(content)
+            script = ''.join(tree.xpath('body/script[contains(text(),"RELATED_PLAYER_ARGS")]/text()'))
             script = re.search(r'\'RELATED_PLAYER_ARGS\':(.*),', script)[1]
             watch_next_response: Dict[str, Any] = json.loads(json.loads(script)['watch_next_response'])
 
@@ -28,38 +91,26 @@ class YoutubeChannel(BaseChannel):
             vid = shareVideoEndpoint['videoId']
             title = shareVideoEndpoint['videoTitle']
             live_url = shareVideoEndpoint['videoShareUrl']
-        elif 'videoDetails' in content:
-            try:
-                script = ''.join(
-                    tree.xpath('//div[@id="player-wrap"]/script[contains(text(),"player_response")]/text()'))
-                script = re.search(r'ytplayer.config = ({.*?});', script)[1]
-                script = json.loads(script)['args']['player_response']
-                videoDetails: Dict[str, Any] = json.loads(script)['videoDetails']
-            except:
-                script = ''.join(tree.xpath('body/script[contains(text(),"ytInitialPlayerResponse")]/text()'))
-                try:
-                    script = re.search(r'var ytInitialPlayerResponse = ({.*});', script)[1]
-                    videoDetails = json.loads(script)['videoDetails']
-                except:
-                    script = re.search(r'window\["ytInitialPlayerResponse"] = (.*?);', script)[1]
-                    videoDetails = json.loads(script)['videoDetails']
+            return LiveCheckResponse(
+                name=self.ch_name,
+                live_status=live_status,
+                title=title,
+                url=live_url,
+                cover=f'https://i.ytimg.com/vi/{vid}/hq720.jpg'
+            )
+        except:
+            return None
 
-            self.ch_name = videoDetails['author']
-            live_status = videoDetails.get('isLive', 0)
-            title = videoDetails['title']
-            vid = videoDetails['videoId']
-            live_url = f'https://youtu.be/{vid}'
-        elif '"isLive":true' not in content:
-            return LiveCheckResponse(name=self.ch_name,
-                                     live_status=0,
-                                     title='',
-                                     url=self.api_url,
-                                     cover=None)
-        else:
-            raise AssertionError('获取直播间信息失败')
-
-        return LiveCheckResponse(name=self.ch_name,
-                                 live_status=live_status,
-                                 title=title,
-                                 url=live_url,
-                                 cover=f'https://i.ytimg.com/vi/{vid}/hq720.jpg')
+    def parse_videoDetails(self, videoDetails: dict):
+        self.ch_name = videoDetails['author']
+        live_status = videoDetails.get('isLive', 0)
+        title = videoDetails['title']
+        vid = videoDetails['videoId']
+        live_url = f'https://youtu.be/{vid}'
+        return LiveCheckResponse(
+            name=self.ch_name,
+            live_status=live_status,
+            title=title,
+            url=live_url,
+            cover=f'https://i.ytimg.com/vi/{vid}/hq720.jpg'
+        )

